components/bpmn/org.wso2.carbon.bpmn.analytics.publisher/src/main/java/org/wso2/carbon/bpmn/analytics/publisher/utils/AnalyticsPublishServiceUtils.java

```diff
@@ -0,0 +1,301 @@
+		                                          AnalyticsPublisherConstants.LAST_PROCESS_INSTANCE_PUBLISH_TIME);
+		if (time == null) {
+			if (instanceQuery.finished().list().size() != 0) {
+				// if the stored time is null in the registry file then send all completed process instances.
+				historicProcessInstanceList =
+						instanceQuery.finished().orderByProcessInstanceStartTime().asc().list();
+			}
+		} else {
+			Date timeInDateFormat = DateConverter.convertStringToDate(time);
+			int listSize = instanceQuery.finished().startedAfter(timeInDateFormat).list().size();
+			if (listSize != 0) {
+				//avoid to return same object repeatedly if the list has only one object
+				if (listSize == 1) {
+					return null;
+				}
+				//send the process instances which were finished after the given date/time in registry
+				historicProcessInstanceList =
+						instanceQuery.finished().startedAfter(timeInDateFormat)
+						             .orderByProcessInstanceStartTime().asc().listPage(1, listSize);
+			}
+		}
+		if (historicProcessInstanceList != null) {
+			log.debug("Write BPMN process instance to the carbon registry...");
+			writePublishTimeToRegistry(historicProcessInstanceList);
+			//return ProcessInstances set as BPMNProcessInstance array
+			return getBPMNProcessInstances(historicProcessInstanceList);
+		}
+		return null;
+	}
+
+	/**
+	 * Get completed task instances which were finished after the given date and time
+	 *
+	 * @return BPMNTaskInstance array if the historic task instance list is not null
+	 */
+	public BPMNTaskInstance[] getCompletedTaskInstances() {
+		HistoryService historyService =
+				BPMNServerHolder.getInstance().getEngine().getHistoryService();
+		HistoricTaskInstanceQuery instanceQuery = historyService.createHistoricTaskInstanceQuery();
+		List<HistoricTaskInstance> historicTaskInstanceList = null;
+		String time = readPublishTimeFromRegistry(AnalyticsPublisherConstants.TASK_RESOURCE_PATH,
+		                                          AnalyticsPublisherConstants.LAST_TASK_INSTANCE_END_TIME);
+		if (time == null) {
+			if (instanceQuery.finished().list().size() != 0) {
+				historicTaskInstanceList =
+						instanceQuery.finished().orderByHistoricTaskInstanceEndTime().asc().list();
+			}
+		} else {
+			Date dateFormat = DateConverter.convertStringToDate(time);
+			int listSize = instanceQuery.finished().taskCompletedAfter(dateFormat).list().size();
+			if (listSize != 0) {
+				//avoid to return same object repeatedly if the list has only one object
+				if (listSize == 1) {
+					return null;
+				}
+				historicTaskInstanceList = instanceQuery.finished().taskCompletedAfter(dateFormat)
+				                                        .orderByHistoricTaskInstanceEndTime().asc()
+				                                        .listPage(1, listSize);
+			}
+		}
+		if (historicTaskInstanceList != null) {
+			log.debug("Write BPMN task instance to the carbon registry...");
```

Comment:
Here also I think we are writing the task end time of the last task, not the task instance. Change the debug message accordingly.